### ISSUE TITLE
Enhancement: [promise-function-async] flag non-`async` functions returning unions with non-`Promise` types

### ISSUE THREAD

**Issue by ericbf** [typescript-eslint/typescript-eslint#6329]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/promise-function-async/

### Description

I am finding cases in my code where it accidentally is returning undefined or some other guard type (null, etc) from a function that we expected to be async, but the rule promise-function-async is not flagging those as I expected it to (allowing us to catch those mistakes).

Please add an option to `promise-function-async` that will allow it to flag any function that returns promises at all, either as the main return type, or as a union containing promises.

### Fail

```typescript
function getSomething(id?: string) {
    if (!id) {
        return undefined
    }

    return db.getByID(id)
}
```

### Pass

```typescript
async function getSomething(id?: string) {
    if (!id) {
        return undefined
    }

    return db.getByID(id)
}
```

### Additional Info

There is an issue for the opposite of my request, #1005. I’d like that feature back (probably behind an option, seeing as it was designated as a bug before).

**Comment by bradzacher** [comment]
I wonder if there's a heuristic we can apply here to avoid the need for an option.

Like "if the function has an explicit or contextual return type that is not `Promise<T>`, then ignore it, else if the implicit return type includes `Promise` at all, report it"

**Comment by ericbf** [comment]
If it has an explicit type that includes non-promises, that would make sense, sure. The real issue, I suppose, is when the implicit type includes a union with non-promises, it should still be flagged, imo.

**Comment by ericbf** [comment]
@bradzacher I put up a PR implementing the heuristic you suggested. Take a look and let me know what you think, or of any suggestions.
